unit/util/expr_initializer.cpp

```diff
@@ -0,0 +1,274 @@
+// Author: Diffblue Ltd.
+
+#include <util/arith_tools.h>
+#include <util/bitvector_types.h>
+#include <util/c_types.h>
+#include <util/expr_initializer.h>
+#include <util/namespace.h>
+#include <util/std_code.h>
+#include <util/symbol_table.h>
+
+#include <testing-utils/use_catch.h>
+
+/// Helper struct to hold useful test components.
+struct expr_initializer_test_environmentt
+{
+  symbol_tablet symbol_table;
+  namespacet ns{symbol_table};
+  source_locationt loc{};
+
+  static expr_initializer_test_environmentt make()
+  {
+    return {};
+  }
+
+private:
+  expr_initializer_test_environmentt() = default;
+};
+
+/// Helper function to create and populate a c_enum_typet type.
+static c_enum_typet make_c_enum_type(
+  const unsignedbv_typet &underlying_type,
+  unsigned int value_count)
+{
+  c_enum_typet enum_type{underlying_type};
+
+  auto &members = enum_type.members();
+  members.reserve(value_count);
+
+  for(unsigned int i = 0; i < value_count; ++i)
+  {
+    c_enum_typet::c_enum_membert member;
+    member.set_identifier("V" + std::to_string(i));
+    member.set_base_name("V" + std::to_string(i));
+    member.set_value(integer2bvrep(i, underlying_type.get_width()));
+    members.push_back(member);
+  }
+  return enum_type;
+}
+
+/// Create a type symbol, a type_tag and a symbol for the tag and populate with
+/// them the symbol_table in the environment.
+static symbolt create_tag_populate_env(
+  const typet &type,
+  expr_initializer_test_environmentt &env)
+{
+  const type_symbolt symbol{"my_type_symbol", type, ID_C};
+  symbolt tag_symbol = [&]() {
+    if(can_cast_type<c_enum_typet>(type))
+    {
+      const c_enum_tag_typet c_enum_tag_type{symbol.name};
+      return symbolt{"my_type_tag_symbol", c_enum_tag_type, ID_C};
```

Comment:
:pick: I may have made similar comments before - however given that this symbol does not define a type, but rather has a type which is a tag variant, then featuring "type_tag" in the name seems misleading to me.

Reply:
Removed as the function is now just returning the `tag_typet`.